keps/sig-autoscaling/0033-20181203-cron-hpa.md

```diff
@@ -0,0 +1,183 @@
+---
+kep-number: 33
+title: Cron Horizontal Pod Autoscaler
+authors:
+  - "@hex108"
+owning-sig: sig-autoscaling
+participating-sigs:
+  - sig-autoscaling
+reviewers:
+  - "@DirectXMan12"
+approvers:
+  - "@DirectXMan12"
+editor: @hex108
+creation-date: 2018-11-27
+last-updated: 2018-11-27
+status: provisional
+---
+
+# Cron Horizontal Pod Autoscaler
+
+Table of Contents
+=================
+
+   * [Cron Horizontal Pod Autoscaler](#cron-horizontal-pod-autoscaler)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories](#user-stories)
+            * [Story 1](#story-1)
+            * [Story 2](#story-2)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [Future work](#future-work)
+
+## Summary
+
+Cron Horizontal Pod Autoscaler(CronHPA) enables us to auto scale workloads(those support `scale`subresource, e.g. deployment, statefulset) periodically using [crontab](https://en.wikipedia.org/wiki/Cron) scheme.
+
+## Motivation
+
+A lot of game players will play games from Friday evening to Sunday evening. It will be better to provide a better experience for players if game servers(pods) could be scaled to a lager size at Friday evening and scaled to normal size at Sunday evening. And that's what game server admins do every week.
+
+Some other customers also do similar things because their products' usage also have peaks and valleys periodically. Time based auto scheduler could scale pods in advance, and will provide a better experience for users. CronHPA is for their requirement.
+
+### Goals
+
+* Users could auto scale their workloads those support `scale` subresource periodically using crontab scheme.
+
+### Non-Goals
+
+* CronHPA does not support workloads that does not support `scale` subresource now.
+
+## Proposal
+
+At first, we'd like to use [CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) to implement this feature. Howerver `CronJob` only supports one `schedule` and users must specify an image and commands, it will be a little inappropriate for above use cases.
+
+We propose a new CRD `CronHPA` which is inspired by `CronJob` (thanks a lot!) to meet users' requirement. Users could specify as many `schedule`s as they want.
+
+`CronHPA` example:
+
+```
+apiVersion: kubernetes.io/v1alpha1
+kind: CronHPA
+metadata:
+  name: example-cron-hpa
+spec:
+  scaleTargetRef:
+    apiVersion: extensions/v1beta1
```

Comment:
Deployment and the other apps are now at `apps/v1` instead of `extensions/v1beta1`. Can you update the references to the current versions?

Reply:
Addressed. Thanks!